cmd/hashers/argon2/calibrate.go

```diff
@@ -0,0 +1,260 @@
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %d iterations\n", currentDuration, config.c.Iterations)
+				}
+
+				if currentDuration > desiredDuration {
+					config.c.Iterations -= 1
+					break
+				}
+				// adjust config
+				config.c.Iterations += 1
+			}
+
+			if verbose {
+				fmt.Fprintf(cmd.ErrOrStderr(), "Decreasing iterations to get under %s:\n", desiredDuration)
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %d iterations\n", currentDuration, config.c.Iterations)
+				}
+
+				// break also when iterations is 1; this catches the case where 1 was only slightly under the desired time and took longer a bit longer on another run
+				if currentDuration < desiredDuration || config.c.Iterations == 1 {
+					break
+				}
+				// adjust config
+				config.c.Iterations -= 1
+			}
+			if verbose {
+				_, _ = resultColor.Fprintf(cmd.ErrOrStderr(), "Settled on %d iterations.\n\n", config.c.Iterations)
+			}
+
+			e := json.NewEncoder(cmd.OutOrStdout())
+			e.SetIndent("", "  ")
+			return e.Encode(config.c)
+		},
+	}
+
+	registerArgon2Flags(cmd.Flags())
+	cmd.Flags().BoolP(FlagVerbose, "v", false, "verbose output")
+	cmd.Flags().IntP(FlagRuns, "r", 2, "runs per probe, median of all runs is taken as the result")
+
+	return cmd
+}
+
+func toKB(b datasize.ByteSize) uint32 {
+	return uint32(b / datasize.KB)
+}
+
+func probe(cmd *cobra.Command, hasher hash.Hasher, runs int, verbose bool) (time.Duration, error) {
+	start := time.Now()
+
+	var mid time.Time
+	for i := 0; i < runs; i++ {
+		mid = time.Now()
+		_, err := hasher.Generate([]byte("password"))
+		if err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Could not generate a hash: %s\n", err)
+			return 0, cmdx.FailSilently(cmd)
+		}
+		if verbose {
+			fmt.Fprintf(cmd.OutOrStdout(), "    took %s in try %d\n", time.Since(mid), i)
+		}
+	}
+
+	return time.Duration(int64(time.Since(start)) / int64(runs)), nil
+}
+
+func registerArgon2Flags(flags *pflag.FlagSet) {
+	flags.StringP(FlagStartMemory, "m", "4GB", "amount of memory to start probing at")
+	flags.String(FlagMaxMemory, "", "maximum memory allowed (default no limit)")
+	flags.String(FlagAdjustMemory, "1GB", "amount by which the memory is adjusted in every step while probing")
+
+	flags.Uint32P(FlagStartIterations, "i", 1, "number of iterations to start probing at")
+
+	flags.Uint8(FlagParallelism, configuration.Argon2DefaultParallelism, "number of threads to use")
+
+	flags.Uint32(FlagSaltLength, configuration.Argon2DefaultSaltLength, "length of the salt in bytes")
+
+	flags.Uint32(FlagKeyLength, configuration.Argon2DefaultKeyLength, "length of the key in bytes")
+}
+
+func configFromFlags(cmd *cobra.Command) (_ *argon2Config, err error) {
```

Comment:
Kinda, it would not be possible for `flagx.GetByteSize` because that has to be a string flag and cobra would not validate that flag on startup. So there would be a panic on malformed values. Do you think we can contribute `ByteSize` to pflag though? That would solve that problem.

Reply:
Or just have `flagx.GetByteSizeE`?

Reply:
sure

Reply:
Ah, there is a way to define custom value flags.